File: ios/Runner/FrameScriptBridge.swift
import Flutter
import WebKit

/// Runs a script inside every frame of the web view, including cross-origin ones.
///
/// The Kakao postcode service draws its result list in an iframe served from
/// `postcode.map.kakao.com`. JavaScript evaluated through
/// `WebViewController.runJavaScript` only ever reaches the main frame, and the
/// same-origin policy keeps that frame's document out of reach from the mirror
/// page, so nothing on the Flutter side can read or click a search result.
///
/// WKWebView's own answer to this is `WKUserScript` with
/// `forMainFrameOnly: false`: the script is evaluated in each frame's own
/// context, where it is same-origin with that frame's document. webview_flutter
/// never exposes the `WKWebViewConfiguration` it builds, so this bridge takes
/// the content controller from the one call the plugin makes on it that we can
/// see — registering the JavaScript channel the Dart side asks for.
final class FrameScriptBridge: NSObject {
  static let shared = FrameScriptBridge()

  private static let channelName = "jikbang/frame_script"

  /// Maps the unique no-op JavaScript channel created by each MirrorPage to
  /// that page's exact content controller. Values are weak so this registry
  /// never extends a WebView's lifetime.
  private let targets = NSMapTable<NSString, WKUserContentController>(
    keyOptions: .strongMemory,
    valueOptions: .weakMemory
  )
  private var installedTargets = Set<String>()

  /// Swaps `addScriptMessageHandler:name:` for our version. Called once, before
  /// any web view exists.
  static func start(with registrar: FlutterPluginRegistrar) {
    WKUserContentController.flrInstallHook()
    let channel = FlutterMethodChannel(
      name: channelName,
      binaryMessenger: registrar.messenger()
    )
    channel.setMethodCallHandler { call, result in
      guard call.method == "setFrameScript" else {
        result(FlutterMethodNotImplemented)
        return
      }
      guard
        let arguments = call.arguments as? [String: Any],
        let source = arguments["script"] as? String,
        let targetChannel = arguments["targetChannel"] as? String
      else {
        result(
          FlutterError(
            code: "bad-arguments",
            message: "setFrameScript expects script and targetChannel.",
            details: nil
          )
        )
        return
      }
      result(shared.install(source: source, targetChannel: targetChannel))
    }
  }

  fileprivate func remember(_ controller: WKUserContentController, channel: String) {
    guard channel.hasPrefix("FrameScriptTarget_") else { return }
    targets.setObject(controller, forKey: channel as NSString)
  }

  /// Adds `source` to the exact content controller identified by the Dart
  /// page's unique channel. Returns whether that controller was available, so the Dart side
  /// can say plainly that automatic selection is off rather than silently
  /// waiting for a click that never comes.
  private func install(source: String, targetChannel: String) -> Bool {
    guard let controller = targets.object(forKey: targetChannel as NSString) else {
      return false
    }
    if installedTargets.contains(targetChannel) { return true }
    controller.addUserScript(
      WKUserScript(
        source: source,
        injectionTime: .atDocumentEnd,
        forMainFrameOnly: false
      )
    )
    installedTargets.insert(targetChannel)
    return true
  }
}

extension WKUserContentController {
  private static var hooked = false

  static func flrInstallHook() {
    guard !hooked else { return }
    hooked = true
    guard
      let original = class_getInstanceMethod(self, #selector(add(_:name:))),
      let replacement = class_getInstanceMethod(self, #selector(flr_add(_:name:)))
    else { return }
    method_exchangeImplementations(original, replacement)
  }

  /// After the exchange this body runs in place of `add(_:name:)`, and the call
  /// to `flr_add` below reaches the original implementation.
  @objc dynamic func flr_add(_ scriptMessageHandler: WKScriptMessageHandler, name: String) {
    FrameScriptBridge.shared.remember(self, channel: name)
    flr_add(scriptMessageHandler, name: name)
  }
}

File: android/app/src/main/kotlin/com/example/jibang_listing_test/FrameScriptBridge.kt
package com.example.jibang_listing_test

import android.webkit.WebView
import androidx.webkit.ScriptHandler
import androidx.webkit.WebViewCompat
import androidx.webkit.WebViewFeature
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugins.webviewflutter.WebViewFlutterAndroidExternalApi
import java.util.WeakHashMap

/**
 * Runs a script inside the Kakao postcode frame, which is cross-origin to the mirror page.
 *
 * The Kakao postcode service draws its result list in an iframe served from
 * `postcode.map.kakao.com`. JavaScript evaluated through
 * `WebViewController.runJavaScript` only ever reaches the main frame, and the
 * same-origin policy keeps that frame's document out of reach from the mirror
 * page, so nothing on the Flutter side can read or click a search result.
 *
 * AndroidX WebKit's answer to this is `WebViewCompat.addDocumentStartJavaScript`:
 * the script is evaluated in every frame whose origin is allowed, in that
 * frame's own context. webview_flutter_android hands out the underlying
 * [WebView] for the identifier the Dart side reads from
 * `AndroidWebViewController.webViewIdentifier`.
 *
 * The iOS counterpart is `ios/Runner/FrameScriptBridge.swift`.
 */
object FrameScriptBridge {
  private const val CHANNEL_NAME = "jikbang/frame_script"

  /**
   * The script each web view carries, so reopening a mirror on a web view that
   * is still alive replaces it rather than stacking a second one. Held weakly:
   * an entry dies with its web view.
   */
  private val installed = WeakHashMap<WebView, ScriptHandler>()

  /** Called once the plugins are registered, before any web view exists. */
  fun start(engine: FlutterEngine) {
    MethodChannel(engine.dartExecutor.binaryMessenger, CHANNEL_NAME)
      .setMethodCallHandler { call, result ->
        if (call.method != "setFrameScript") {
          result.notImplemented()
          return@setMethodCallHandler
        }
        val source = call.argument<String>("script")
        val webViewId = call.argument<Number>("webView")?.toLong()
        val origins = call.argument<List<String>>("origins")
        if (source == null || webViewId == null || origins.isNullOrEmpty()) {
          result.error(
            "bad-arguments",
            "setFrameScript expects script, webView and origins.",
            null,
          )
          return@setMethodCallHandler
        }
        try {
          result.success(install(engine, webViewId, source, origins.toSet()))
        } catch (error: IllegalArgumentException) {
          // An origin rule WebView does not accept.
          result.error("bad-arguments", error.message, null)
        }
      }
  }

  /**
   * Returns whether the script went in, so the Dart side can say plainly that
   * automatic selection is off (a System WebView too old for document-start
   * scripts, or a web view that is already gone) rather than silently waiting
   * for a click that never comes.
   */
  private fun install(
    engine: FlutterEngine,
    webViewId: Long,
    source: String,
    origins: Set<String>,
  ): Boolean {
    if (!WebViewFeature.isFeatureSupported(WebViewFeature.DOCUMENT_START_SCRIPT)) {
      return false
    }
    val webView = WebViewFlutterAndroidExternalApi.getWebView(engine, webViewId) ?: return false
    installed.remove(webView)?.remove()
    installed[webView] = WebViewCompat.addDocumentStartJavaScript(webView, source, origins)
    return true
  }
}
